src/Tmds.Systemd/ServiceManager.cs

```diff
@@ -33,5 +37,57 @@ public static string InvocationId
                 return _invocationId;
             }
         }
+
+        private static bool CheckServiceManager()
+        {
+            // No point in testing anything unless it's Unix
+            if (Environment.OSVersion.Platform != PlatformID.Unix)
+            {
+                return false;
+            }
+
+            // We've got invocation id, it's systemd >= 232 running a unit
+            if (InvocationId != null)
+            {
+                return true;
+            }
+
+            // Either it's not a service, or systemd is < 232, do a bit more digging
+            try
+            {
+                // Test parent process
+                var parentPid = getppid();
+                var pidString = parentPid.ToString(NumberFormatInfo.InvariantInfo);
+
+                // If parent PID is not 1, this may be a user unit, in this case it must match MANAGERPID envvar
+                if (parentPid != 1
+                    && Environment.GetEnvironmentVariable("MANAGERPID") != pidString)
+                {
+                    return false;
+                }
+
+                // Check process name for the parent process to match "systemd\n"
+                using (var commFile = File.Open("/proc/" + pidString + "/comm", FileMode.Open, FileAccess.Read, FileShare.Read))
```

Comment:
Simplify this to `File.ReadAllText("/proc/" + pidString + "/comm") == "systemd"`?

Reply:
My original code had almost the same code (except it had the ASCII encoding explicitly specified to prevent the need for its detection). But, since the /proc is a virtual FS, any file there may have arbitrary size, and you can't check the file length to match since the virtual FS metadata always has zero length:
```
~$ ls -la /proc/1/comm
-rw-r--r-- 1 root root 0 Jun 27 15:02 /proc/1/comm
```

So I moved to this per-byte reading to prevent potential (unlikely, yes, but possible) hangs and/or memory overflows if /proc stops playing well.

Reply:
> But, since the /proc is a virtual FS, any file there may have arbitrary size, and you can't check the file length to match since the virtual FS metadata always has zero length:

Do you mean File.ReadAllText doesn't work?

File.ReadAllBytes should work: https://github.com/dotnet/corefx/blob/40364e539572e9dad9c8a2eb165fc9af28e5664a/src/System.IO.FileSystem/src/System/IO/File.cs#L340-L342

> So I moved to this per-byte reading to prevent potential (unlikely, yes, but possible) hangs and/or memory overflows if /proc stops playing well.

I think it's safe to assume proc keeps on doing what it does. Or do you have a specific case in mind?

> per-byte reading

I think FileStream may be buffering, so to the OS it will be a single read call.

Reply:
It does work, but has too much "magic" inside. I agree ReadAllBytes is a viable alternative if I turn my paranoia down a few notches. I'll take a look at it later today.

Reply:
Moved to `File.ReadAllBytes` and `Span<byte>.SequenceEqual()` instead of hard-coding array of chars for better readability.